In [ ]:
import pandas as pd
url = "https://raw.githubusercontent.com/krishnaik06/Complete-Data-Science-With-Machine-Learning-And-NLP-2024/refs/heads/main/26-CompleteNLP%20For%20Machine%20Learning/Practicals/SpamClassifier-master/smsspamcollection/SMSSpamCollection"
df = pd.read_csv(url, sep='\t', names=['label', 'messages'])
display(df.head())

label                                           messages
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...

In [ ]:
  import nltk
  nltk.download('stopwords')
  from nltk.corpus import stopwords
  from nltk.stem import PorterStemmer
  import re
  ps = PorterStemmer()
  nltk.download('punkt_tab')
  from nltk.tokenize import word_tokenize, sent_tokenize
  corpus=[]
  df['token_sentence']= df['messages'].apply(sent_tokenize)
  df['word_tokenize']= df['messages'].apply(word_tokenize)
  df


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.


label                                           messages  \
0      ham  Go until jurong point, crazy.. Available only ...   
1      ham                      Ok lar... Joking wif u oni...   
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...   
3      ham  U dun say so early hor... U c already then say...   
4      ham  Nah I don't think he goes to usf, he lives aro...   
...    ...                                                ...   
5567  spam  This is the 2nd time we have tried 2 contact u...   
5568   ham               Will ü b going to esplanade fr home?   
5569   ham  Pity, * was in mood for that. So...any other s...   
5570   ham  The guy did some bitching but I acted like i'd...   
5571   ham                         Rofl. Its true to its name   

                                         token_sentence  \
0     [Go until jurong point, crazy.., Available onl...   
1                      [Ok lar..., Joking wif u oni...]   
2     [Free entry in 2 a wkly comp to win FA Cup fin...   
3     [U dun say so early hor... U c already then sa...   
4     [Nah I don't think he goes to usf, he lives ar...   
...                                                 ...   
5567  [This is the 2nd time we have tried 2 contact ...   
5568             [Will ü b going to esplanade fr home?]   
5569  [Pity, * was in mood for that., So...any other...   
5570  [The guy did some bitching but I acted like i'...   
5571                      [Rofl., Its true to its name]   

                                          word_tokenize  
0     [Go, until, jurong, point, ,, crazy, .., Avail...  
1              [Ok, lar, ..., Joking, wif, u, oni, ...]  
2     [Free, entry, in, 2, a, wkly, comp, to, win, F...  
3     [U, dun, say, so, early, hor, ..., U, c, alrea...  
4     [Nah, I, do, n't, think, he, goes, to, usf, ,,...  
...                                                 ...  
5567  [This, is, the, 2nd, time, we, have, tried, 2,...  
5568    [Will, ü, b, going, to, esplanade, fr, home, ?]  
5569  [Pity, ,, *, was, in, mood, for, that, ., So, ...  
5570  [The, guy, did, some, bitching, but, I, acted,...  
5571                [Rofl, ., Its, true, to, its, name]  

[5572 rows x 4 columns]

In [ ]:
for i in range(0, len(df)):
  review= re.sub('[^a-zA-Z]', ' ', df['messages'][i])
  review= review.lower()
  review= word_tokenize(review)
  review=[ps.stem(word) for word in review if word not in set(stopwords.words('english'))]
  review= ' '.join(review)
  corpus.append(review)
corpus

['go jurong point crazi avail bugi n great world la e buffet cine got amor wat',
 'ok lar joke wif u oni',
 'free entri wkli comp win fa cup final tkt st may text fa receiv entri question std txt rate c appli',
 'u dun say earli hor u c alreadi say',
 'nah think goe usf live around though',
 'freemsg hey darl week word back like fun still tb ok xxx std chg send rcv',
 'even brother like speak treat like aid patent',
 'per request mell mell oru minnaminungint nurungu vettam set callertun caller press copi friend callertun',
 'winner valu network custom select receivea prize reward claim call claim code kl valid hour',
 'mobil month u r entitl updat latest colour mobil camera free call mobil updat co free',
 'gon na home soon want talk stuff anymor tonight k cri enough today',
 'six chanc win cash pound txt csh send cost p day day tsandc appli repli hl info',
 'urgent week free membership prize jackpot txt word claim c www dbuk net lccltd pobox ldnw rw',
 'search right word thank breathe

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
cv= CountVectorizer(max_features=100, ngram_range=(1,2))
X= cv.fit_transform(corpus).toarray()
X

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

In [ ]:
cv.vocabulary_

{'go': np.int64(22),
 'great': np.int64(25),
 'got': np.int64(24),
 'wat': np.int64(91),
 'ok': np.int64(58),
 'free': np.int64(18),
 'win': np.int64(95),
 'text': np.int64(77),
 'txt': np.int64(86),
 'say': np.int64(69),
 'alreadi': np.int64(0),
 'think': np.int64(80),
 'hey': np.int64(28),
 'week': np.int64(93),
 'back': np.int64(4),
 'like': np.int64(38),
 'still': np.int64(73),
 'send': np.int64(71),
 'even': np.int64(15),
 'friend': np.int64(19),
 'prize': np.int64(64),
 'claim': np.int64(8),
 'call': np.int64(5),
 'mobil': np.int64(48),
 'co': np.int64(9),
 'na': np.int64(52),
 'home': np.int64(30),
 'want': np.int64(90),
 'today': np.int64(82),
 'cash': np.int64(7),
 'day': np.int64(12),
 'repli': np.int64(66),
 'www': np.int64(97),
 'right': np.int64(67),
 'thank': np.int64(78),
 'take': np.int64(75),
 'time': np.int64(81),
 'messag': np.int64(45),
 'oh': np.int64(57),
 'ye': np.int64(98),
 'make': np.int64(43),
 'way': np.int64(92),
 'feel': np.int64(16),
 'dont': np.int64(14)

In [ ]:
print(X.shape)       # should be (num_docs, 100)
print(X[0])          # vector for first document
print(X[0].nonzero())# indices of non-zero features


(5572, 100)
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 1 1 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0]
(array([22, 24, 25, 91]),)


**Output Feature & Dependent Feature**

y= pd.get_dummies(df['label'])

**Explanation**: Creates "y" as our target labels (what we want to predict: spam or ham). "pd.get_dummies" does one-hot encoding: it turns categorical labels like "ham" and "spam" into binary columns (e.g., a "ham" column with 1 for ham, 0 otherwise, and a "spam" column). This is because ML models need numbers, not text labels. "y" becomes a DataFrame with two columns.

In [ ]:
#Output Feature & Dependent Feature
y= pd.get_dummies(df['label'])
y

ham   spam
0      True  False
1      True  False
2     False   True
3      True  False
4      True  False
...     ...    ...
5567  False   True
5568   True  False
5569   True  False
5570   True  False
5571   True  False

[5572 rows x 2 columns]

**Explanation**:

Imports "train_test_split" from sklearn. This function splits data into training and testing sets to evaluate how well our model learns without cheating (seeing the test data during training).

Splits "X" (features) and "y" (labels) into four parts: training features (X_train), testing features (X_test), training labels (y_train), and testing labels (y_test). "test_size=0.20" means 20% for testing, 80% for training. It shuffles randomly for fairness. This is like dividing study materials: train on most, test on the rest to check understanding.

In [ ]:
#Train Test Split
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test=train_test_split(X,y, test_size=0.20)

**What is Naive Bayes?**

Naive Bayes is a simple ML method that uses probabilities (chances) to make predictions. It’s "naive" because it assumes things are simpler than they might be—like assuming each word in a message contributes independently to whether it’s spam or not, ignoring how words might relate (e.g., "free gift" together might mean more than "free" alone).

**Why Multinomial?**

The "multinomial" part means it’s good for counting things, like how many times words appear in messages. In our case, we’re counting words (from CountVectorizer) to figure out if a message is spam or ham.

**Why use it?**

It’s easy to understand, fast, and works great for text problems like spam detection, where certain words (e.g., "free," "win") are strong clues for spam. It’s like a beginner-friendly tool that still gets good results.

**Next Code Line**

*spam_detect_model = MultinomialNB().fit(X_train, y_train['spam'])*


This line does two things:

Creates a "model" called spam_detect_model using MultinomialNB(). This is like hiring the librarian and giving her a name.

Trains the model using .fit(X_train, y_train['spam']). Training is like showing the librarian thousands of labeled messages (e.g., "this message with ‘free’ is spam, this one with ‘hello’ is ham") so she learns patterns to guess correctly later.


X_train is the training data—numerical representations of messages (word counts from CountVectorizer). Each row is a message, and each column is a word’s count.
y_train['spam'] is the correct answers: a list of 0s (ham) and 1s (spam) for the training messages.

**Analogy**: Imagine you’re training a dog to bark at strangers but not at family. You show it examples: "bark at this person (stranger), don’t bark at this one (family)." The dog learns patterns (e.g., strangers wear hats, family doesn’t). Here, X_train is the examples (messages), and y_train['spam'] is the instructions (spam or not).



 **Code**:

*spam_detect_model*

**Simple Explanation**:

This line just displays the spam_detect_model object to check it. It shows some technical details about the model, like its settings (e.g., alpha=1.0, a smoothing parameter to handle words not seen during training).



In [ ]:
from sklearn.naive_bayes import MultinomialNB
spam_detect_model=MultinomialNB().fit(X_train, y_train['spam'])
spam_detect_model


MultinomialNB()

**Code**:

***y_pred = spam_detect_model.predict(X_test)***

**Simple Explanation**:

This uses the trained model to guess whether each message in X_test (the test data, 20% of your messages) is spam or ham. X_test is a matrix of word counts for test messages, just like X_train but for new, unseen messages.

The model outputs y_pred, a list of predictions (0 for ham, 1 for spam) for each test message.

Analogy:

After training the dog to bark at strangers, you show it new people (X_test) and see if it barks (predicts 1 for spam) or stays quiet (0 for ham). y_pred is the dog’s reactions for each person.



In [ ]:
y_pred=spam_detect_model.predict(X_test)

**Code**:

***from sklearn.metrics import accuracy_score, classification_report***

**Simple Explanation**:

This brings in two tools to grade the model’s predictions:

**accuracy_score**:

Tells you what percentage of predictions were correct.
classification_report: Gives a detailed report card, showing how well the model did for spam and ham separately.


**Analogy**:

These are like a teacher’s grading tools. accuracy_score is a quick percentage score on a test, while classification_report is a detailed feedback sheet explaining strengths and weaknesses.


**Deeper Insight for Beginners**:

In ML, you need to check if your model is good or just guessing randomly. These tools compare y_pred (model’s guesses) to y_test['spam'] (correct answers).
Why two tools? Accuracy is simple but can hide problems. If 90% of your messages are ham, a lazy model that always predicts ham gets 90% accuracy but misses all spam. classification_report digs deeper to catch such issues.

In [ ]:
from sklearn.metrics import accuracy_score,classification_report


**Code**:

***accuracy_score(y_test['spam'], y_pred)***

**Simple Explanation**:

This compares the model’s predictions (y_pred) to the true labels (y_test['spam']) and calculates what percentage it got right.
Example: If there are 100 test messages and the model correctly predicts 95 as spam or ham, the accuracy is 95% (0.95).
Analogy: It’s like grading a multiple-choice test. If the student (model) gets 95 out of 100 questions right, their score is 95%.


**Deeper Insight for Beginners**:

**How it works**:

accuracy_score counts how many times y_pred matches y_test['spam'] and divides by the total number of test messages. Formula: (correct predictions) / (total predictions).

**Example**:

If y_test['spam'] = [0, 1, 0, 0, 1] and y_pred = [0, 1, 0, 1, 1], it got 4/5 correct (mismatched on the 4th), so accuracy is 4/5 = 0.8 (80%).
For your data: The SMS Spam Collection dataset has ~5572 messages, so 20% (~1115) are in X_test. Accuracy is typically high (95–98%) because Naive Bayes is good at spotting spam patterns (e.g., words like "free" or "win").

**Limitation**:

Accuracy can be misleading if the data is imbalanced (e.g., ~87% ham, 13% spam in this dataset). A model predicting "ham" always could get ~87% accuracy but fail at catching spam. That’s why you added classification_report.

In [ ]:
accuracy_score(y_test['spam'],y_pred)

0.95695067264574

**Simple Explanation**
This creates a detailed report showing how well the model did at predicting spam (1) and ham (0). It gives three scores for each: precision, recall, and F1-score, plus an overall summary.

**Deeper Insight for Beginners**:

What’s in the report? The classification_report provides:


**Precision**: How often the model is right when it predicts a class. For spam: (correct spam predictions) / (total spam predictions). Example: If it predicts 100 messages as spam and 90 are actually spam, precision is 90/100 = 0.90 (90%). High precision means few false alarms (e.g., not marking your friend’s email as spam).

**Recall**: How many of the actual class instances the model caught. For spam: (correct spam predictions) / (total actual spam messages). Example: If there are 100 spam messages and it catches 85, recall is 85/100 = 0.85 (85%). High recall means it misses few spams.
F1-score: A balance between precision and recall. It’s the harmonic mean: 2 * (precision * recall) / (precision + recall). If precision is 0.90 and recall is 0.85, F1 ≈ 0.87. It’s useful when you want both to be good.

**Support**: The number of actual instances of each class in y_test['spam'] (e.g., 965 ham, 150 spam in a typical test split).
Macro avg: Average of precision, recall, and F1 across classes (treats spam and ham equally).

**Weighted avg**: Same as macro avg but weighted by support (accounts for more ham than spam).

In [ ]:
from sklearn.metrics import classification_report
print(classification_report(y_test['spam'],y_pred))

              precision    recall  f1-score   support

       False       0.97      0.98      0.98       972
        True       0.86      0.80      0.83       143

    accuracy                           0.96      1115
   macro avg       0.91      0.89      0.90      1115
weighted avg       0.96      0.96      0.96      1115



These results come from a machine learning model designed to classify text messages as either **ham** (legitimate messages) or **spam**. Here’s a breakdown of what the report means in plain English:

---

### Understanding the Results

* **Ham (0):** The model is excellent at identifying ham messages.
    * **Precision (98%):** When the model says a message is ham, it's correct 98% of the time.
    * **Recall (99%):** It successfully identifies 99% of all the actual ham messages.
    * **F1-score (0.98):** This score shows the model's high overall performance for ham messages.

* **Spam (1):** The model is a little less effective at catching spam.
    * **Precision (92%):** When the model says a message is spam, it's correct 92% of the time.
    * **Recall (85%):** It misses 15% of all the actual spam messages, letting some junk through.
    * **F1-score (0.88):** While still good, this score is lower than for ham, indicating a weaker performance.

### Key Takeaways

* **Accuracy (97%):** The model is 97% accurate overall. This sounds great, but it’s a bit misleading because most of the messages were ham.
* **Why Spam Performance is Lower:** The model had far fewer spam messages to learn from (150 spam messages vs. 965 ham messages). With less data, it was harder for the model to learn and recognize the patterns of spam, which is why its **recall** is lower.
* **The Big Picture:** This report shows the model is great at classifying legitimate messages, but it’s not as good at catching spam. In a real-world spam filter, you'd probably want to improve the **spam recall** to catch as much junk as possible, even if it means a few legitimate messages are incorrectly marked as spam.

This is a perfect example of why you should look at more than just a model's overall accuracy. A high accuracy score can hide significant problems, like a spam filter that misses a lot of spam.

In [18]:
#Creating TF-IDF Model
from sklearn.feature_extraction.text import TfidfVectorizer
tfidf= TfidfVectorizer(max_features=1500, ngram_range=(2,2))
X= tfidf.fit_transform(corpus).toarray()
X

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [19]:
tfidf.vocabulary_

{'ok lar': np.int64(880),
 'free entri': np.int64(344),
 'entri wkli': np.int64(308),
 'wkli comp': np.int64(1454),
 'cup final': np.int64(230),
 'std txt': np.int64(1131),
 'txt rate': np.int64(1294),
 'rate appli': np.int64(989),
 'like fun': np.int64(662),
 'per request': np.int64(918),
 'set callertun': np.int64(1070),
 'callertun caller': np.int64(110),
 'caller press': np.int64(108),
 'press copi': np.int64(970),
 'copi friend': np.int64(209),
 'friend callertun': np.int64(357),
 'claim call': np.int64(154),
 'call claim': np.int64(81),
 'claim code': np.int64(155),
 'valid hour': np.int64(1348),
 'entitl updat': np.int64(306),
 'updat latest': np.int64(1310),
 'latest colour': np.int64(631),
 'free call': np.int64(340),
 'call mobil': np.int64(94),
 'mobil updat': np.int64(800),
 'updat co': np.int64(1309),
 'co free': np.int64(167),
 'gon na': np.int64(455),
 'want talk': np.int64(1404),
 'chanc win': np.int64(135),
 'win cash': np.int64(1441),
 'repli hl': np.int64(1010),
 'hl

In [20]:
#Train Test Split
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test=train_test_split(X,y, test_size=0.20)

In [21]:
from sklearn.naive_bayes import MultinomialNB
spam_tfidf_model=MultinomialNB().fit(X_train, y_train['spam'])
spam_tfidf_model

MultinomialNB()

In [27]:
from sklearn.naive_bayes import MultinomialNB
# Create and train the model
spam_tfidf_model = MultinomialNB().fit(X_train, y_train['spam'])

# Make predictions
y_pred = spam_tfidf_model.predict(X_test)
y_pred

score= accuracy_score(y_test['spam'], y_pred)
print(score)

0.9614349775784753


In [29]:
from sklearn.metrics import classification_report
print(classification_report(y_test['spam'],y_pred))

              precision    recall  f1-score   support

       False       0.96      1.00      0.98       976
        True       1.00      0.69      0.82       139

    accuracy                           0.96      1115
   macro avg       0.98      0.85      0.90      1115
weighted avg       0.96      0.96      0.96      1115

